### PR TITLE
Updated Prettify/Uglify process to remove need to use control characters

### PR DESCRIPTION
**Pull request by dhedey**

### CODE REVIEW

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs:100]: I'm torn between arguing that we should use chevrons, because there's already a git pattern of using chevrons for markers like this.

And arguing that this is fine, and we should actively *avoid* using chevrons, to ensure that we never tread on git's toes.

Thoughts?

File: AdfDataflowFilePrettifier/regenerateTestFiles.sh
Comment (Brondahl) [inline AdfDataflowFilePrettifier/regenerateTestFiles.sh:1]: Seems odd to actively use the `-fromStdIn` mode, when you achieve that my `cat`ing the path? Any reason you didn't just use the `-fromFile` mode?

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs:91]: I suspect that you've lost this subtlety? Will double check.

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs]: Can you make this a constructor that accepts just the two `IndexOf()` return values.

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs:74]: Can you add something here that also checks for stray whitespace in the script definition?
One quick option is probably to just validate it as a JSON string?
https://stackoverflow.com/a/14977915/1662268

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs:128]: Can you wrap `GetNextBlockIndices` into `GetAllBlockIndices`, which returns a (potentially empty) collection of non-null Blocks.
Then foreach over those valid blocks, handling `if(!.Any())` upfront by simply returning the input string as-is.

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs]: see below => `foreach`

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs]: I can't decide whether to request further `ReadToStartOfBlock` and `ReadBlockContents` methods because they'll add clarity (and shorten the line), or whether they'd just be pointless extra layers of methds.

So, I'll suggest the concept and leave it to you to decide :D

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [reply AdfDataflowFilePrettifier/Prettifier.cs:74]: Although ... feel free to declare that that's outside the scope of this change, and leave me to do it if I want it.

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [reply AdfDataflowFilePrettifier/Prettifier.cs:91]: Yes, you have.

Hmm ... so ... this is a specific thing to support `kdiff` because `kdiff` is a bit bad at line endings.
You can pick your resultant line endings, but you can't end up with mixed LEs. :(
So these `"\\r\\n"` strings get turned into `"\r\n"` by the prettifier, and then turned into `"\n"` by kdiff, and then turned into `"\\n"` by the uglifier. But `"\\r\\n"` != `"\\n"`. [citation needed]

I should have created a test for that case :(

I do still need that code ... but it shouldn't be hidden away pretending to be "part of core prettification/uglification". It should be an explicit step in the process, layered separately.

Do you have any suggestions *other* than adding some sort of magic indicator (ASCII control character or otherwise)?

**Review by Brondahl — CHANGES_REQUESTED** [review]
A few refactoring points, but one major functional issue, alas.

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs:100]: I'm not sure I feel strongly either way, so by default would keep as is. As you say, I think chevrons would require some testing :)

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs:91]: Ouch. This seems like a kdiff issue! But actually, more generally, isn't there an issue with kdiff trampling the line endings and if you make any manual changes in kdiff? Like resolving any merges manually?

Azure (I would hope) would only ever use one line ending anyway, so wouldn't the best thing to do be to just be sneaky and replace all line endings with the azure ones when we uglify? Or alternatively boot up kdiff with a \r\n line ending style http://kdiff3.sourceforge.net/doc/options.html using the cs command line flag http://kdiff3.sourceforge.net/doc/documentation.html ? 

Also - we should imo tell git attributes to turn autocrlf off for these json files if it's not already, in case that impacts this process at all? (although existing checkouts with autocrlf on might have issues).

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs:128]: I like this idea a lot 👍

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs]: Mmmm I can...

But I view this as a record more than a class (although my VS said I wasn't using a high enough version of C# to use records). In general, I'm a fan of keeping data and code separate, and I don't see why moving the code to compute the end of block markers away from the rest of the code and into the class helps particularly?

Happy to if you'd rather...

File: AdfDataflowFilePrettifier/regenerateTestFiles.sh
Comment (dhedey) [reply AdfDataflowFilePrettifier/regenerateTestFiles.sh:1]: My thinking was that if we're printing to stdOut then it feels more symmetric to read from stdIn, particularly if we're in a bash script. 👍 

But either way was fine, and it was actually easier to understand how to use -fromStdIn than -fromFile because the program is very strict about the order of command line args haha.

Happy to change if you'd rather

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs]: I think it's easier to read / check the code if it's all the index use is together in the same method, so I think it makes sense to leave it there :)

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs:91]: There's also an argument that we don't need the output of prettifying and uglifying to neccessarily be opposites - the process can simply normalize the output, as long as it's still valid ADF code?

Ie the mergetool could also clean up line endings and that would be okay?

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs:74]: I haven't got time to change this right now, but happy to do so when as part of v2 markups

**Review by dhedey — COMMENTED** [review]
Thanks for the review! Added a few thoughts / questions for you :)

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [reply AdfDataflowFilePrettifier/Prettifier.cs:100]: Fair. Let's leave it as is.

File: AdfDataflowFilePrettifier/regenerateTestFiles.sh
Comment (Brondahl) [reply AdfDataflowFilePrettifier/regenerateTestFiles.sh:1]: I'm not particularly fussed :)

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs:127]: Yes, this feels much easier to follow, to me.

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [reply AdfDataflowFilePrettifier/Prettifier.cs:91]: So ...

> Ouch. This seems like a kdiff issue!

Yes, agreed. But fixing kdiff seems like a much less viable solution :(

> Azure (I would hope) would only ever use one line ending anyway

Yeah, you'd hope so, right? But no. The majority of Endings in the scripted bits are LF, but it occasionally has some CRLF with no obvious rules.
But IIRC one time I tried "fixing" it, and the UI reverted it back to CRLFs :(

>There's also an argument that we don't need the output of prettifying and uglifying to neccessarily be opposites - the process can simply normalize the output, as long as it's still valid ADF code?
>Ie the mergetool could also clean up line endings and that would be okay?

I'm really keen to avoid anything that introduces LineEnding churn in the files. This tool is already introducing a bunch of surprise and mild confusion. I really don't want to add to that by having it cause spontaneous changes (which will be hard to understand since many diff tools explain EOL changes poorly) to lines that weren't different in either parent commit.

>so wouldn't the best thing to do be to just be sneaky and replace all line endings with the azure ones when we uglify?

In essence that's what my old code did. The problem is "the azure ones" is not a predictable value. See above.
So my old code's solution was to make everything in pretty-land consistently LFs, but mark some of those LFs as extra-double-special, and mapped those back to CRLFs when uglifying.

>Or alternatively boot up kdiff with a \r\n

Would be fine if Azure were *consistently* annoying ... but see above: it isn't. :(

>Also - we should imo tell git attributes to turn autocrlf off for these json files if it's not already, in case that impacts this process at all? (although existing checkouts with autocrlf on might have issues).

Uhh ... 
IMO autocrlf *should* always be off - it shouldn't be git's job to decide what the checked-out / pushed contents of the file should look like.
However ... if a project has decided it wants autocrlf *on* ... then ... ??? ... this tool shouldn't override that, I think?
Since this tool isn't interacting with checkout or push (which AFAIU are the only ops that autocrlf affects?) I think it shouldn't care? The tool should just not change any line endings anywhere.

File: AdfDataflowFilePrettifier/AdfDataflowFilePrettifier.Tests/TestFiles/SampleDataFlow3_WithEscapedCRInDefinition_Pretty.txt
Comment (Brondahl) [inline AdfDataflowFilePrettifier/AdfDataflowFilePrettifier.Tests/TestFiles/SampleDataFlow3_WithEscapedCRInDefinition_Pretty.txt:130]: FYI, this is the location of an example CRLF in the script definition.
The ugly file contains: `"== 1 &&\r\ntrue())"`

Comment (Brondahl) [reply AdfDataflowFilePrettifier/AdfDataflowFilePrettifier.Tests/TestFiles/SampleDataFlow3_WithEscapedCRInDefinition_Pretty.txt:130]: My *guess* would be that the rule is "if there's a new line in the text that defines a operation expression that was entered by the user, then that new line is (or *can* be) a CRLF"? 

Recall that, overall we have:
* A JSON file, containing a prop that is a string
* That represents a script file, containing a sequence of blocks
* Which represent configured operations of a dataset
* Where some of that configuration consists of multi-line free-text entered by the user.

=:0

Comment (Brondahl) [reply AdfDataflowFilePrettifier/AdfDataflowFilePrettifier.Tests/TestFiles/SampleDataFlow3_WithEscapedCRInDefinition_Pretty.txt:130]: eesh ... it's even worse :D ... there's an extra bullet point:

* where that free text could legitimately be defining multi-line string values!

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [reply AdfDataflowFilePrettifier/Prettifier.cs]: I'm probably not that fussed either way. Leave it as is.

Not sure I could articulate why I would do it that way, myself.
I'm generally on-board with data classes which do some of the set up in their constructor. 🤷 
Thinking about it further, I suspect it's a habit derived from too many instances of noticing that data class has fewer dependencies that props, and the logic to derive the props from the dependencies is duplicated in multiple locations where we need to create the data object.

Clearly not a concern here, though :D

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (dhedey) [inline AdfDataflowFilePrettifier/Prettifier.cs:176]: This is my suggestion for replacing `\\r`. It avoids use of control characters to make things more explicit for someone doing the merge / diff.

`\r\n` appears very rarely, so this shouldn't affect the general readability of the script, but it also makes things a lot less magic, and gives a person doing manual merges the ability to add in `\r` if they wish to.

There's a tiny risk that a `[[CRLF]]` might already be present in the file, or that someone might accidentally edit it to eg `[[CRLF]` or something and it to remain in the file. If we'd want to be extra careful and willing to add in a tiny bit more magic, we could add a control character in the middle of the string to reduce the likelihood of false positives.

Also, I chose [[CRLF]] instead of [[CR]] because the former is more clearly an end of line marker and more intuitive (even if it's not strictly correct).

Interested to hear your thoughts.

File: AdfDataflowFilePrettifier/AdfDataflowFilePrettifier.Tests/TestFiles/SampleDataFlow3_WithEscapedCRInDefinition_InvalidJson.txt
Comment (Brondahl) [inline AdfDataflowFilePrettifier/AdfDataflowFilePrettifier.Tests/TestFiles/SampleDataFlow3_WithEscapedCRInDefinition_InvalidJson.txt]: The filename should now be **UN**escaped CR :D

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Prettifier.cs:51]: Nice, thank you.

File: AdfDataflowFilePrettifier/Program.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/Program.cs:47]: :+1:

File: AdfDataflowFilePrettifier/AdfDataflowFilePrettifier.Tests/PrettificationTests.cs
Comment (Brondahl) [inline AdfDataflowFilePrettifier/AdfDataflowFilePrettifier.Tests/PrettificationTests.cs:78]: :+1: thanks

File: AdfDataflowFilePrettifier/Prettifier.cs
Comment (Brondahl) [reply AdfDataflowFilePrettifier/Prettifier.cs:176]: Yes, I think that's definitely the right solution.

Thanks for the explanation on CR/CRLF - that was the first thought I had.
Are you concerned that people who would otherwise recognise `[[CRLF]]` might be interpret `[[CR]]` as something else (due to lack of prompt to consider it a line ending?)

If so that seems reasonable.

Comment (Brondahl) [reply AdfDataflowFilePrettifier/Prettifier.cs:176]: Other than that, I'm happy with this. Thank you!

Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs:176]: > Are you concerned that people who would otherwise recognise [[CRLF]] might be interpret [[CR]] as something else (due to lack of prompt to consider it a line ending?)

Yes, exactly. If you weren't aware of what was going on, `[[CR]]` doesn't immediately say to me "line ending weirdness", and could conceivably be something in the ADF scripting world.

Would, say, `[[CarriageReturn]]` be preferred to `[[CRLF]]`? I think I have a slight preference for `[[CRLF]]`, but up to you?

Let me know. Otherwise, I think this is ready to merge if/when you're happy. Thanks for all your help with reviewing this! 😀

Comment (dhedey) [reply AdfDataflowFilePrettifier/Prettifier.cs:176]: We could also just leave it as `\\r` of course, although I think `[[CRLF]]` is a slight preference. Still, very happy to give you a deciding vote :D

Comment (Brondahl) [reply AdfDataflowFilePrettifier/Prettifier.cs:176]: ooh ... leaving it as `\\r` is kinda tempting.
But I think I prefer `[[CRLF]]` since it's clearer that it's a marker we've put in deliberately (and not, say, a bug in the prettifier).

So I think we're good to leave it as it currently is.

**Comment by Brondahl** [comment]
I think we've now completed all the discussions and it's ready to be merged in. Do you agree?

**Comment by dhedey** [comment]
Agreed 👍🏼

**Pull request merged**
